gstreamer: 1.16.2 -> 1.18.0

nh2 opened the pull request:
###### Motivation for this change

https://gstreamer.freedesktop.org/releases/1.18/

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Note that building docs is disabled with this upgrade; somebody needs to package [`hotdoc`](https://hotdoc.github.io/) for that to work again.

Feel free to push fixes straight into my branch, because I only did this work to try something out with `srt`, I do not actually have a lot of time to work on gstreamer upgrades currently.

nh2 committed:
gstreamer: 1.16.2 -> 1.18.0

Fixes #98769.

Important changes from https://gstreamer.freedesktop.org/releases/1.18/:

* `gst-validate` was renamed to `gst-devtools` upstream:

    > * the `gst-validate` tarball has been superseded by
    >   the `gst-devtools` tarball for consistency with the git module name.

* `gst-python` is now Python 3 only:

    > * Python 2.x is no longer supported

nh2 commented:
@GrahamcOfBorg build gst_all_1.gst-editing-services gst_all_1.gst-plugins-bad gst_all_1.gst-plugins-good gst_all_1.gst-rtsp-server gst_all_1.gst-validate gst_all_1.gst-libav gst_all_1.gst-plugins-base gst_all_1.gst-plugins-ugly gst_all_1.gst-vaapi gst_all_1.gstreamer

ajs124 commented:
cc @doronbehar #98769

nh2 committed:
gst_all_1.gst-plugins-bad: Add upstream patch to fix SRT usage

worldofpeace commented:
Quick question, now that we don't have ancient gstreamer in top-level could we move everything out of `gst_all_1`?

nh2 commented:
> Quick question, now that we don't have ancient gstreamer in top-level could we move everything out of `gst_all_1`?

I'd be in favour of that, either making all packages top-level, or renaming `gst_all_1` to `gstreamer_all`, or maybe even better,  both (so that you can conveniently build build the entire `gstreamer` attrset with an easy-to-remember name).

doronbehar commented:
I would support this as well. IIRC I once wanted to patch a package to use gst v1.0+ plugins and the default attributes in the top level were for the deprecated packages (now no longer in Nixpkgs), and it took me some time to realize why won't it work.

nh2 commented:
> Quick question, now that we don't have ancient gstreamer in top-level could we move everything out of `gst_all_1`?

Just for clarification, I think this should be done in a separate PR independent from version upgrades (and I don't have the bandwidth for it).

nh2 commented:
@GrahamcOfBorg build

nh2 commented:
@GrahamcOfBorg build gst_all_1 fractal pitivi

nh2 commented:
This should be more finished now.

`gst_all_1` and `pitivi` now build fine when based on `master`, and I just switched the PR to `staging`.

jtojnar commented on pkgs/development/libraries/gstreamer/bad/default.nix:
```diff
@@ -87,25 +89,27 @@ let
     })
-    ./fix_pkgconfig_includedir.patch
-    # https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/merge_requests/1235
-    ./opencv-4.3.patch
```
Remove this file?

nh2 replied:
Done.

jtojnar commented on pkgs/applications/video/pitivi/default.nix:
```diff
@@ -127,7 +127,7 @@ in python3Packages.buildPythonApplication rec {
     gst-libav
-    gst-validate
+    gst-devtools
   ]);
```
gst-transcoder is now part of gstreamer so it will likely conflict. There are also some upstream patches that might be relevant. Or maybe update pitivi.

nh2 replied:
It hasn't conflicted so far, but independent of it, I cannot start `pitivi`: It fails with the probably unrelated error:

```
Traceback (most recent call last):
  File "/nix/store/0qk27xn3za106wpg4l3miiqgfqhdk9s4-pitivi-0.999/bin/..pitivi-wrapped-wrapped", line 169, in <module>
    _run_pitivi()
  File "/nix/store/0qk27xn3za106wpg4l3miiqgfqhdk9s4-pitivi-0.999/bin/..pitivi-wrapped-wrapped", line 146, in _run_pitivi
    from pitivi import application
  File "/nix/store/0qk27xn3za106wpg4l3miiqgfqhdk9s4-pitivi-0.999/lib/pitivi/python/pitivi/application.py", line 34, in <module>
    from pitivi.mainwindow import MainWindow
  File "/nix/store/0qk27xn3za106wpg4l3miiqgfqhdk9s4-pitivi-0.999/lib/pitivi/python/pitivi/mainwindow.py", line 42, in <module>
    from pitivi.medialibrary import AssetThumbnail
  File "/nix/store/0qk27xn3za106wpg4l3miiqgfqhdk9s4-pitivi-0.999/lib/pitivi/python/pitivi/medialibrary.py", line 173, in <module>
    class AssetThumbnail(Loggable):
  File "/nix/store/0qk27xn3za106wpg4l3miiqgfqhdk9s4-pitivi-0.999/lib/pitivi/python/pitivi/medialibrary.py", line 189, in AssetThumbnail
    EMBLEMS[status].append(GdkPixbuf.Pixbuf.new_from_file_at_size(
gi.repository.GLib.Error: gdk-pixbuf-error-quark: Couldn’t recognize the image file format for file “/nix/store/0qk27xn3za106wpg4l3miiqgfqhdk9s4-pitivi-0.999/share/pitivi/pixmaps/asset-proxied.svg” (3)
```

jtojnar replied:
Sounds like pitivi needs `strictDeps = false;`. I thought I already added that :woman_shrugging:

nh2 replied:
@jtojnar I also just figured that out, doing it in PR #101398 (and I rebased this PR on top of that commit).

jtojnar commented:
We should also update gst-python, notably

> - Python 2.x is no longer supported

Regarding hotdoc, we should remove gtk:

> * API and plugin documentation are no longer built with `gtk_doc` . The `gtk_doc` documentation has been removed in favour of a [new unified documentation module](https://gitlab.freedesktop.org/gstreamer/gst-docs) built with `hotdoc` (also see "Documentation improvements" section below). Distributors should use the [documentation release tarball](https://gstreamer.freedesktop.org/data/src/gstreamer-docs/) instead of trying to package `hotdoc` and building the documentation from scratch.

Next

> * gst-plugins-bad now includes an internal copy of [libusrsctp](https://github.com/sctplab/usrsctp), as there are problems in `usrsctp` with global shared state, lack of API stability guarantees, and the absence of any kind of release process. We also can't rely on distros shipping a version with the fixes we need. Both firefox and Chrome bundle their own copies too. It is still possible to build against an external copy of `usrsctp` if so desired.

sctp is disabled so that does not affect us.

> * `nvcodec` no longer needs the NVIDIA NVDEC/NVENC SDKs available at build time, only at runtime. This allows distributions to ship this plugin by default and it will just start to work when the required run-time SDK libraries are installed by the user, without users needing to build and install the plugin from source.

We are disabling nvdec & nvenc at the moment; figuring out how to load them at runtime would be painful so let’s go on with it.

> * the `gst-editing-services` tarball is now named `gst-editing-services` for consistency (used to be `gstreamer-editing-services` ).

not packaged.

> * the `gst-validate` tarball has been superseded by the `gst-devtools` tarball for consistency with the git module name.

:heavy_check_mark:

nh2 committed:
pitivi: Use `gst-transcoder` from `gst-plugins-bad`

nh2 committed:
pitivi: 0.999 -> 2020.09.1

jtojnar commented:
What’s up with the `fixDarwinDylibNames` commits?

nh2 committed:
pitivi: Fix .svg loading at startup, see #56943 #89691

nh2 commented:
> What’s up with the `fixDarwinDylibNames` commits?

@jtojnar Not sure what you mean, perhaps you got an email from Github in between I rebased onto staging and changed the github merge base?

nh2 commented:
On `staging`, the builds fail due to dependency `ghostscript` failing:

```
./base/fapi_ft.c:1909:1: error: control reaches end of non-void function [-Werror=return-type]
 1909 | }
      | ^
cc1: some warnings being treated as errors
```

I have a branch [`gstreamer-1.18.0-nonstaging`](https://github.com/nh2/nixpkgs/tree/gstreamer-1.18.0-nonstaging) that one can use to test non-staging.

> > Sounds like pitivi needs `strictDeps = false;`. I thought I already added that woman_shrugging
>
>  I also just figured that out, doing it in PR #101398 (and I rebased this PR on top of that commit).

With these fixes, pitivi starts fine on the new gstreamer.

nh2 commented:
> Regarding hotdoc, we should remove gtk:

@jtojnar I think I already did that everywhere, removing:

```diff
-    gtk-doc
-    docbook_xsl
-    docbook_xml_dtd_43
```

> We are disabling nvdec & nvenc at the moment; figuring out how to load them at runtime would be painful so let’s go on with it.

Yes, I would like to eventually get that working to be able to encode h264 very efficiently on nvidia, so if you have an idea for it, please share :)

> > * the `gst-editing-services` tarball is now named `gst-editing-services` for consistency (used to be `gstreamer-editing-services` ).
> 
> not packaged.

It is, [pkgs/development/libraries/gstreamer/ges/default.nix](https://github.com/NixOS/nixpkgs/blob/b4349c13a6d0ae0e82269332effe4b6a56572840/pkgs/development/libraries/gstreamer/ges/default.nix). But I have updated it.

jtojnar reviewed (approved):
Pitivi seems to work fine when I rebase this branch + the `strictDeps` patch onto unstable.

Build logs look fine too, except for unknown meson flags `examples` in `ugly`, and `nvdec` and `nvenc` in `bad`.

nh2 commented:
> Build logs look fine too, except for unknown meson flags `examples` in `ugly`, and `nvdec` and `nvenc` in `bad`.

Fixed.

doronbehar commented on pkgs/development/libraries/gstreamer/core/default.nix:
```diff
@@ -16,34 +15,28 @@
+  outputs = [
+    "out"
+    "dev"
+    # "devdoc" # disabled until `hotdoc` is packaged in nixpkgs
```
```suggestion
    # "devdoc" # disabled until `hotdoc` is packaged in nixpkgs, see:
    # - https://github.com/NixOS/nixpkgs/pull/98767
    # - https://github.com/NixOS/nixpkgs/issues/98769#issuecomment-702296551
```

jtojnar replied:
I wonder if it even makes sense to try to build docs ourselves instead of using a separate package just extracting the tarball, as suggested in the release notes https://github.com/NixOS/nixpkgs/pull/99345#issuecomment-714186302

doronbehar commented on pkgs/development/libraries/gstreamer/libav/default.nix:
```diff
@@ -37,6 +37,15 @@ stdenv.mkDerivation rec {
 
+  mesonFlags = [
+    "-Ddoc=disabled" # `hotdoc` not packaged in nixpkgs as of writing
+    ];
```
```suggestion
  ];
```

doronbehar reviewed (commented):
Just a few formatting issues, otherwise LGTM. Thanks for working on this.

nh2 commented:
I've run `nixpkgs-review` for a while on `gstreamer-1.18.0-nonstaging`; there are various failures there but I couldn't find any that are related to gstreamer.

Merging.

nh2 commented:
@samueldr found that this broke `aarch64`: https://hydra.nixos.org/build/130355045/nixlog/42

```
Has header "bcm_host.h" : NO 

sys/rpicamsrc/meson.build:30:4: ERROR: Problem encountered: Could not find bcm_host.h. Please pass the location of this header via -Drpi-header-dir=/path
```

I had tried to use ofborg to figure out whether the upgrade breaks other arches, but I couldn't tell, because staging was apparently broken on on aarch64 at the time:

https://github.com/NixOS/nixpkgs/pull/99345/checks?check_run_id=1300702358

and I don't have an aarch64 device on which I could test it well.

**I think our CI story for this isn't good yet.** I think ofborg should also test staging PRs against `nixpkgs-unstable` to be able to tell quickly whether they would work at all. Otherwise it'll always be a blind shot.

nh2 commented:
It broke Darwin as well: #104062
